Content.Shared/Tools/Systems/ToolRefinableSystem.cs

```diff
@@ -1,55 +1,274 @@
+    private void OnDoAfter(Entity<ToolRefinableComponent> ent, ref WelderRefineDoAfterEvent args)
     {
-        if (args.Cancelled)
+        if (args.Cancelled || args.Used == null)
+            return;
+
+        var (uid, component) = ent;
+
+        var getIsBlocked = new GetIsToolRefineBlockedEvent(args.Used.Value, component.QualityNeeded);
+        RaiseLocalEvent(ref getIsBlocked);
+        if (!getIsBlocked.IsRefinable)
+        {
+            _popup.PopupPredicted(getIsBlocked.BlockCause, args.User, args.User);
             return;
+        }
+
+        PopupOnRefine(uid, component, args.Used.Value, args.User);
+
+        // TODO: Use RandomPredicted https://github.com/space-wizards/RobustToolbox/pull/5849
+        var rndSeed = SharedRandomExtensions.HashCodeCombine((int)_gameTiming.CurTick.Value, args.User.Id, uid.Id);
+        var rng = new System.Random(rndSeed);
+
+        var spawns = EntitySpawnCollection.GetSpawns(component.RefineResult, rng);
+        List<EntityUid> spawned = new(spawns.Count);
+        foreach (var protoId in spawns)
+        {
+            var sliceUid = PredictedSpawnNextToOrDrop(protoId, uid);
+            spawned.Add(sliceUid);
+
+            if (!_container.IsEntityOrParentInContainer(sliceUid))
+            {
+                var randVect = rng.NextPolarVector2(2.0f, 2.5f);
+                _physics.SetLinearVelocity(sliceUid, randVect);
+            }
+        }
+
+        if (TryComp<ToolRefinableSolutionComponent>(uid, out var comp))
+        {
+            TryGetSourceSolutionForTransfer(uid, comp.SolutionToSplit, out var solutionInfo);
 
-        if (_net.IsClient)
+            foreach (var spawnedUid in spawned)
+            {
+                // Fills new slice if original entity allows.
+                if (solutionInfo.HasValue && comp.SolutionToSet != null)
+                {
+                    var (sourceSoln, sourceSolution) = solutionInfo.Value;
+                    var sliceVolume = sourceSolution.Volume / FixedPoint2.New(spawns.Count);
+
+                    var lostSolution = _solutionContainer.SplitSolution(sourceSoln, sliceVolume);
+                    FillResult(spawnedUid, comp.SolutionToSet, lostSolution);
+                }
+            }
+        }
+
+        if (component.Sound != null)
+            _audio.PlayPredicted(component.Sound, Transform(uid).Coordinates, args.User, AudioParams.Default.WithVolume(-2));
+
+        var ev = new BeforeToolRefineFinishedEvent(args.User);
+        RaiseLocalEvent(uid, ref ev);
+
+        if(ev.Cancelled)
             return;
 
-        var xform = Transform(uid);
-        var spawns = EntitySpawnCollection.GetSpawns(component.RefineResult, _random);
-        foreach (var spawn in spawns)
+        _body.GibBody(uid);
+        _destructible.DestroyEntity(uid);
+    }
+
+    private void PopupOnRefine(EntityUid uid, ToolRefinableComponent component, EntityUid used, EntityUid user)
```

Comment:
Typically shouldn't separate them when they can tuple.

```suggestion
    private void PopupOnRefine(Entity<ToolRefinableComponent> ent, EntityUid used, EntityUid user)
```